SI-6331 deconst If type / refine equality of floating point Constant types.

retronym opened the pull request:
Review by @paulp

/cc @xeno-by 
1. I'd suggest you remove `ConstantType#isNaN` from `scala.reflect.api.Constants#ConstantApi`.
2. Do you expect this output when compiling a macro in partest?

```
quick.partest:
[scalacfork] Compiling 1 file to /Users/jason/code/scala/build/quick/classes/partest
[scalacfork] // produced from source-/Users/jason/code/scala/src/partest/scala/tools/partest/package.scala,line-98,offset=2789
[scalacfork] object Test extends App {
[scalacfork]   val $u: c.universe.type = c.universe;
[scalacfork]   import $u._
[scalacfork]   val tpe = evidence$1.in($m).tpe
[scalacfork]   println(tpe)
[scalacfork] }
[scalacfork] // produced from source-/Users/jason/code/scala/src/partest/scala/tools/partest/package.scala,line-98,offset=2789
[scalacfork] object Test extends App {
[scalacfork]   val $u: c.universe.type = c.universe;
[scalacfork]   val $m: $u.Mirror = c.universe.rootMirror;
[scalacfork]   import $u._, $m._
[scalacfork]   val tree = Block(List(Apply(Select(Select(build.This(staticPackage("scala").asModule.moduleClass), newTermName("Predef")), newTermName("println")), List(Apply(Select(Apply(Select(Select(build.Ident(staticPackage("scala")), newTermName("StringContext")), newTermName("apply")), List(Literal(Constant("trace> ")), Literal(Constant("\\nres: ")), Literal(Constant(" = ")), Literal(Constant("\\n")))), newTermName("s")), List(exprCode.in($m).tree, exprType.in($m).tree, a.in($m).tree))))), a.in($m).tree)
[scalacfork]   println(tree)
[scalacfork] }
[propertyfile] Updating property file: /Users/jason/code/scala/build/quick/classes/partest/partest.properties
[stopwatch] [quick.partest.timer: 10.808 sec]
```

retronym committed:
SI-6331 Avoid typing an If tree with a constant type.

The fast path in typedIf added in 8552740b avoided lubbing the if/else branch types
if they are identical, but this fails to deconst the type. This could lead to the entire
if expression being replaced by a constant.

Also introduces a new tool in partest for nicer checkfiles.

// in Test.scala
trace(if (t) -0d else 0d)

// in Test.check
trace> if (Test.this.t)
  -0.0
else
  0.0
res: Double = -0.0

retronym committed:
Refine equality of Constant types over floating point values.

The constant types for 0d and -0d should not be equal.

This is implemented by checking equality of the result of
doubleToRawLongBits / floatToRawIntBits, which also correctly
considers two NaNs of the same flavour to be equal.

Followup to SI-6331.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/382/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1089/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/382/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1089/

paulp commented on src/partest/scala/tools/partest/package.scala:
```diff
@@ -73,4 +73,31 @@ package object partest {
```
I'm thinking scala/tools/nsc/package.scala.  Why punish ourselves by hiding it in partest.

retronym replied:
That might take an expert hand to wire it into the fast track mechanism. Let's do that separately from this PR.

xeno-by replied:
There's a very good example of integrating a "normal" macro into fast track:
1) macro (you already did this step): https://github.com/scala/scala/blob/2.10.x/src/reflect/scala/reflect/runtime/package.scala
2) registration in definitions: https://github.com/scala/scala/blob/2.10.x/src/reflect/scala/reflect/internal/Definitions.scala#L489
3) bridge in fast track: https://github.com/scala/scala/blob/2.10.x/src/compiler/scala/tools/reflect/FastTrack.scala#L37

Nothing scary. Just two extra lines and a redeployment of starr.

retronym commented:
PLS REBUILD ALL  :pray:

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/383/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1090/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1090/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/383/

retronym commented:
In order to clean up `Constant#hashCode`, I need to understand `Constant#tag`. It was added to `Constant#equals` way back when in 4720d45a83 without adding it to `hashCode` (EDIT: which doesn't break the contract); `Constant#tag` itself was added even longer ago in the dark days of CVS 4ae459ef.

My question: Is there an example where `this.value == other.value && this.tag != other.tag`? If not, why is `tag` considered in `equals`? Penny for your thoughts, @odersky?

retronym committed:
Test for consistency of Constant#{equals, hashCode}.

For the examples I've constructed, they are consistent, but
I put this down to good luck, rather than good management.
The next commit will address this.

retronym committed:
Improve Constant#hashCode

 - Incorporate `tag`, which is considered by equals, to reduce collisions.
 - Use the result of floatToRawIntBits(value) / doubleToRawLongBits(value),
    rather than value. This wasn't strictly necessary as (0d.## == (-0d).##)
    but this is more obviously correct.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/388/

retronym commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1095/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1095/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/388/

gkossakowski commented:
What's the status of this one? Shall commit breaking build be squashed with ones fixing it? Do we need to wait for @odersky's feedback?

retronym commented:
All the build failures above were spurious (the build didn't even start).

All commits in the sequence actually build. I forgot to amend `hashCode`, but it still worked because `(-0d).## == 0d.##`. I added a test to show this, then went and made `hashCode` more obviously correct, and also better distributed to boot.

In short: I don't think we need to squash, and I'm happy with all the commits. The open question I had ("why does equals bother to compare `tag`) is noted in a comment.

jsuereth commented:
Yeah, I think this one is good to go as is, just with follow on work.

jsuereth commented:
@paulp or @xeno-by  can I get a LGTM here?

paulp commented:
What kind of nutbag would find something with this kind of test coverage to look anything less than good.  LGTM.

You sound like you're ready to lead some religious revivals.

```
"Can I get an LGTM?"
"LGTM!"
"Can I get a hallelujah, LGTM!"
"Hallelujah, LGTM!"
"Can I get a test case?"
"Not as a general rule, no."
```
